app/backend/chat_history/cosmosdb.py

```diff
@@ -0,0 +1,198 @@
+import os
+import time
+from typing import Any, Dict, Union
+
+from azure.cosmos.aio import ContainerProxy, CosmosClient
+from azure.identity.aio import AzureDeveloperCliCredential, ManagedIdentityCredential
+from quart import Blueprint, current_app, jsonify, request
+
+from config import (
+    CONFIG_CHAT_HISTORY_COSMOS_ENABLED,
+    CONFIG_COSMOS_HISTORY_CLIENT,
+    CONFIG_COSMOS_HISTORY_CONTAINER,
+    CONFIG_CREDENTIAL,
+)
+from decorators import authenticated
+from error import error_response
+
+chat_history_cosmosdb_bp = Blueprint("chat_history_cosmos", __name__, static_folder="static")
+
+
+@chat_history_cosmosdb_bp.post("/chat_history")
+@authenticated
+async def post_chat_history(auth_claims: Dict[str, Any]):
+    if not current_app.config[CONFIG_CHAT_HISTORY_COSMOS_ENABLED]:
+        return jsonify({"error": "Chat history not enabled"}), 405
+
+    container: ContainerProxy = current_app.config[CONFIG_COSMOS_HISTORY_CONTAINER]
+    if not container:
+        return jsonify({"error": "Chat history not enabled"}), 405
+
+    entra_oid = auth_claims.get("oid")
+    if not entra_oid:
+        return jsonify({"error": "User OID not found"}), 401
+
+    try:
+        request_json = await request.get_json()
+        id = request_json.get("id")
+        answers = request_json.get("answers")
+        title = answers[0][0][:50] + "..." if len(answers[0][0]) > 50 else answers[0][0]
+        timestamp = int(time.time() * 1000)
+
+        await container.upsert_item(
+            {"id": id, "entra_oid": entra_oid, "title": title, "answers": answers, "timestamp": timestamp}
+        )
+
+        return jsonify({}), 201
+    except Exception as error:
+        return error_response(error, "/chat_history")
+
+
+@chat_history_cosmosdb_bp.post("/chat_history/items")
+@authenticated
+async def get_chat_history(auth_claims: Dict[str, Any]):
+    if not current_app.config[CONFIG_CHAT_HISTORY_COSMOS_ENABLED]:
+        return jsonify({"error": "Chat history not enabled"}), 405
+
+    container: ContainerProxy = current_app.config[CONFIG_COSMOS_HISTORY_CONTAINER]
+    if not container:
+        return jsonify({"error": "Chat history not enabled"}), 405
+
+    entra_oid = auth_claims.get("oid")
+    if not entra_oid:
+        return jsonify({"error": "User OID not found"}), 401
+
+    try:
+        request_json = await request.get_json()
+        count = request_json.get("count", 20)
+        continuation_token = request_json.get("continuation_token")
+
+        res = container.query_items(
+            query="SELECT c.id, c.entra_oid, c.title, c.timestamp FROM c WHERE c.entra_oid = @entra_oid ORDER BY c.timestamp DESC",
+            parameters=[dict(name="@entra_oid", value=entra_oid)],
+            max_item_count=count,
+        )
+
+        # set the continuation token for the next page
+        pager = res.by_page(continuation_token)
+
+        # Get the first page, and the continuation token
+        try:
+            page = await pager.__anext__()
```

Comment:
Consider iterating over the by_page result directly and avoiding explicit calls to await pager.__anext__(), i.e. process each page as soon as it’s available without awaiting when there are no more pages.

Reply:
I think you mean to just do:
async for page in pager:
  async for item in page:

As we use similar code for other Azure Python SDKs elsewhere in this repo. That wouldn't give us the continuation token, right, as that would exhaust all the pages?

Reply:
Yes, but you should still be able to get the continuation token. Something like (untested):

```python
        items = []
        async for page in container.query_items(
            query="SELECT c.id, c.entra_oid, c.title, c.timestamp FROM c WHERE c.entra_oid = @entra_oid ORDER BY c.timestamp DESC",
            parameters=[{"name": "@entra_oid", "value": entra_oid}],
            max_item_count=count,
        ).by_page(continuation_token):
            async for item in page:
                items.append(
                    {
                        "id": item.get("id"),
                        "entra_oid": item.get("entra_oid"),
                        "title": item.get("title", "untitled"),
                        "timestamp": item.get("timestamp"),
                    }
                )
            # Update continuation token after processing the page
            continuation_token = page.continuation_token if hasattr(page, "continuation_token") else None

            # Break if no continuation token (i.e., last page)
            if not continuation_token:
                break
```

Just a suggestion. I think its fine as it is :-)